controlplane/kubeadm/controllers/upgrade.go

```diff
@@ -82,6 +82,26 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		}
 	}
 
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration != nil {
+		if err := workloadCluster.UpdateAPIServerInKubeadmConfigMap(ctx, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.APIServer); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to update api server in the kubeadm config map")
+		}
+
+		if err := workloadCluster.UpdateControllerManagerInKubeadmConfigMap(ctx, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to update controller manager in the kubeadm config map")
+		}
+
+		if err := workloadCluster.UpdateSchedulerInKubeadmConfigMap(ctx, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.Scheduler); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to update scheduler in the kubeadm config map")
+		}
```

Comment:
I'm not sure this PR is the right place for this comment, but IMO we should refactor the code handling the ConfigMap changes in order to compose a list of transformations to apply to the ConfigMap, and then, if there are transformations, read/update the config map only once (instead of reading/upgrading many times)

Reply:
I agree, we can refactor to apply the changes lazily. I'll create an issue for it.

Reply:
Filed an [issue](https://github.com/kubernetes-sigs/cluster-api/issues/4007) for this.

Reply:
Fine for me, Thanks!